### PR TITLE
Allow the nginx-ingress-controller to be configured by a configmap

### PR DESCRIPTION
**Pull request by axinojolais**
This resolves https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/186

### CODE REVIEW

File: cluster/juju/layers/kubernetes-worker/templates/ingress-replication-controller.yaml [hunk 1]
```diff
@@ -45,3 +45,4 @@ spec:
         args:
         - /nginx-ingress-controller
         - --default-backend-service=$(POD_NAMESPACE)/default-http-backend
+        - --nginx-configmap=$(POD_NAMESPACE)/nginx-load-balancer-conf
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-worker/templates/ingress-replication-controller.yaml]: Thanks for the contribution!  I see you are adding a configmap flag here. How are you going to get the Nginx configuration to the worker?

Comment (axinojolais) [inline cluster/juju/layers/kubernetes-worker/templates/ingress-replication-controller.yaml]: Easy : you just create a ConfigMap object. For example :
```
apiVersion: v1
data:
  body-size: 1024m
kind: ConfigMap
metadata:
  name: nginx-load-balancer-conf
```

You can see the list of all the options you can tweak at https://github.com/kubernetes/contrib/blob/master/ingress/controllers/nginx/configuration.md#allowed-parameters-in-configuration-configmap

Using the ConfigMap allows to deploy a "generic" ingress controller that anyone can tweak to their own needs later on.

**Comment by mbruzek** [comment]
I have not yet deployed this change, but it looks fine to me. I would like to add information in the README about how to configure or how a user would configure this option.

**Comment by chuckbutler** [comment]
This change LGTM. The default behavior didn't appear to change when the configmap was missing, and I like how easy this makes it tuning the ingress bits.

More to @mbruzek's point though we should at bare minimum land a simple change to the readme along with this so it's clear to end users that this is a configurable option. I'm not going to block this PR on that, however. 

This also needs to be piggybacked into github.com/kubernetes/kubernetes as our source destination has moved from here to there as our large divergent branch has been caught up.

With this being a single file change it should be trivial to do so. If you don't mind @axinojolais would you point this PR at the upstream destination? If not, I can certainly carry this forward on your behalf by pulling this into a branch based on tip and denote that you've made the contribution. (they love squashed commits upstream)

+1 LGTM barring some minor doc updates.

**Comment by axinojolais** [comment]
Will update the README and retarget

**Comment by chuckbutler** [comment]
Do we know if axino was able to retarget this against upstream before heading out on vacation?

I tried grepping the pull request list but didn't see it... however that might have just been me (considering its 7:30 in the morning and i have no coffee)

**Comment by Cynerva** [comment]
Doesn't look like it. @chuckbutler :coffee:

**Comment by mbruzek** [comment]
This PR needs to be rebased to the upstream repository. For it to land.

I checked our addons directories and I don't see a file named "nginx-load-balancer-conf" and I worry about adding this configuration will cause a file not found error somewhere. @axinojolais gave me an "easy" example of what to put there, but without the file in this same PR I am not comfortable merging this.

```

File: cluster/juju/layers/kubernetes-worker/templates/ingress-replication-controller.yaml (hunk 1 above)
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-worker/templates/ingress-replication-controller.yaml]: The above link 404's for me so I can't see the options.

Comment (axinojolais) [inline cluster/juju/layers/kubernetes-worker/templates/ingress-replication-controller.yaml]: @mbruzek https://github.com/kubernetes/ingress/blob/master/controllers/nginx/configuration.md#allowed-parameters-in-configuration-configmap

**Comment by axinojolais** [comment]
@mbruzek FWIW, the official example from Google does have the --nginx-configmap option : https://github.com/kubernetes/ingress/blob/master/examples/customization/configuration-snippets/nginx/nginx-ingress-controller.yaml

I think it works just fine if the file doesn't exist.

**Comment by mbruzek** [comment]
In the lined example the nginx configuration definition file exists in the directory so the ConfigMap object would exist in that example cluster. I was confused because in this PR, the ConfigMap is not defined but the ingress controller is configured to use a ConfigMap **if/when** it exists. Meaning the users would have to add their own ConfigMap object (with a specific name) to the Kubernetes system after install.

If the users are expected to add their own custom ConfigMap after install is it not reasonable to expect them to configure the ingress controller to point to that object? If this PR added the `nginx-load-balancer-conf.yaml` file to the template directory the charm would render of both these objects for the users when the charm re-renders the templates on charm upgrade or other hook events that might re-render the templates.

I spoke with @axinojolais on IRC and he confirmed that this PR would be setting the ingress controller to use an undefined object. If that is the only way to do it I am OK with this approach it just seems unusual to me and I am worried about

**Comment by axinojolais** [comment]
This is deprecated, closing.

**Pull request closed without merging**